### DISCUSSION TITLE
How to properly set eslint with Nextjs 11+ in a monorepo

### DISCUSSION THREAD

**Discussion by KnowYourselves** [vercel/next.js#28725] [Help]
Hello!

I'm having a little problem setting up the default eslint config in a monorepo.  I have the following structure:
```
MyApp/
├─ express-backend/
│  ├─ ...
├─ nextjs-frontend/
│  ├─ .eslintrc.json
│  ├─ ...
├─ docker-compose.yml
```

Where `exrpess-backend` is an express aplication and `nextjs-frontend` is my next app created via `create-next-app` as per the documentation. 

My problem is that following the instruccions and doing `yarn lint` in the next app generates the config file, but it doesn't work at all. I suppose that's because it isn't the source folder, but I've tried many different values for the `rootDir` option but nothings works.

Doing yarn eslint --init works fine although the next plugin for eslint doesn't work either in that case.  I'd just like to understand how to properly use the next plugin for eslint in this use case.

Thanks!

**Comment by belgattitude** [comment]
Generally I create a base [.eslint.base.js](https://github.com/belgattitude/nextjs-monorepo-example/blob/main/.eslintrc.base.js) with a common configuration that is valid for all projects. I tend to use this as a "safe" base:

```
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended',
    'plugin:jest/recommended',
    'plugin:sonarjs/recommended',
``` 

Then I extend in packages ans apps. ie: [.eslintrc.js](https://github.com/belgattitude/nextjs-monorepo-example/blob/main/apps/web-app/.eslintrc.js)

ie:

```
.
├── apps
│   ├── express
│   │   ├── package.json
│   │   └── .eslintrc.js  (extends base + specific to express)      
│   └── nextjs           
│       ├── package.json
│       └── .eslintrc.js (extends base + specific to nextjs/jsx-a11y/...)
│
├── packages
│   ├── db-main-prisma          (extends base, specific to prisma)
│    │   ├── package.json
│    │   └── .eslintrc.js
│    └── ui-lib                 
│       ├── package.json
│       └── .eslintrc.js (extends base + specific to react/react-hooks/jsx a11y...)
├ .eslint.base.js
└ package.json
```

For example in a nextjs app I would extend the base with things like

```js
module.exports = {
  root: true,
  ignorePatterns: ['node_modules/*', '.next'],
  extends: [
    '../../.eslintrc.base.js',
    // Add specific rules for react and nextjs
    'plugin:react/recommended',
    'plugin:react-hooks/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:@next/next/recommended',
    'plugin:testing-library/react',
  ],
  //...
}
```

Of course you can create multiple bases if you feel. 

I don't remember having problems with it. The only thing sometimes in github actions, I already had a linter step and to avoid linting twice, I just disable nextjs eslint, see [here](https://github.com/belgattitude/nextjs-monorepo-example/blob/ac6840c8165c3698b2e129527896a83f7f66f59c/.github/workflows/ci-web-app.yml#L96). 

But with recent advances in nextjs (eslint cache), I'll probably lint in the nextjs build step.

**Reply by KnowYourselves** [reply]
Thanks! Something I don't understand though, Is it necesary for that to work for the father folder to have a package.json?

EDIT: I've given a look to the repo and I can't make the base config work, as soon as I import it eslint breaks in the base projects

**Reply by belgattitude** [reply]
> Something I don't understand though, Is it necesary for that to work for the father folder to have a package.json?

I don't think so, but I never tried. I always have one (workspace enabled).

>  can't make the base config work, as soon as I import it eslint breaks in the base projects

Without the error message, I can't really guess :smile: 



I've updated the folder structure in the comment above... maybe we have some differences

**Reply by KnowYourselves** [reply]
That's the weird thing, as soon as I extend any of the .eslint files of my apps using the base it just stops working, the same if I try to use the next plugin in the eslint and there's just no error message in the output for eslint in VSCode... that's why it's weird 🤔

**Reply by belgattitude** [reply]
Mmm weird. Are you sure you're not using a locally installed eslint ? Like an older version ? Check also that all eslint plug-ins are present in devdeps .. all eslintrc are with the same format. Js / json.

Can't remember having issues with extend. 

Good luck
